src/mesh/domain.hpp

```diff
@@ -0,0 +1,172 @@
+  //  also contains a label for defining which region the index shape is assigned too 
+  class IndexShape {
+    private:
+      std::array<IndexRange,NDIM> x_;
+      std::array<int,NDIM> entire_ncells_;
+
+    public:
+  
+      IndexShape() {};
+
+      IndexShape(const int & nx1, const int & nx2, const int & nx3, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2,nx3}, ng) {};
+
+      IndexShape(const int & nx1, const int & nx2, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2}, ng) {};
+
+      IndexShape(const int & nx1, const int & ng) 
+      : IndexShape( std::vector<int> {nx1}, ng) {};
+
+      IndexShape(const std::vector<int> & interior_dims, const int & ng) { 
+        assert(interior_dims.size()<=NDIM && "IndexShape cannot be initialized, the number of "
+              "dimensions exceeds the statically set dimensions, you will need to change the NDIM "
+              "constant.");
+        for( int dim=1, index=0; dim<=NDIM; ++dim, ++index){
+          if (dim <= interior_dims.size()) {
+            assert( interior_dims.at(index) > 0 && "IndexShape cannot be initialized with fewer "
+                "than 1 interior cells for each dimension");
+            x_[index].start = ng;
+            x_[index].end = x_[index].start + interior_dims.at(index) - 1;
+            entire_ncells_[index] = interior_dims.at(index) + 2*ng;
+          } else {
+            x_[index].start = 0;
+            x_[index].end = 0;
+            entire_ncells_[index] = 1;
+          }
+        }
+      };
+
+      inline int x1s(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? 0 : x_[0].start; }
+      
+      inline int x2s(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[1].start; }
+      
+      inline int x3s(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[2].start; }
+      
+      inline int x1e(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0]-1 : x_[0].end; }
+      
+      inline int x2e(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1]-1 : x_[1].end; }
+      
+      inline int x3e(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2]-1 : x_[2].end; }
+
+      inline int nx1(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0] : x_[0].ncells(); }
+      
+      inline int nx2(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1] : x_[1].ncells(); }
+      
+      inline int nx3(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2] : x_[2].ncells(); }
+
+
+      void GetIndices(const IndexShapeType & type, 
```

Comment:
👍

Reply:
So this was discussed, the problem with the GetIndices method as it is, is that it is easy to mess up the order.

Reply:
That is an issue, yes.